src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -47,72 +50,123 @@ public static StreamingClientProvider getStreamingClientProviderInstance() {
     return StreamingClientProviderSingleton.streamingClientProvider;
   }
 
+  /**
+   * Builds the loading cache to register streaming clients
+   *
+   * @param streamingClientHandler The handler to create clients with
+   * @return A loading cache to register clients
+   */
+  public static LoadingCache<Map<String, String>, SnowflakeStreamingIngestClient> buildLoadingCache(
+      StreamingClientHandler streamingClientHandler) {
+    return Caffeine.newBuilder()
+        .maximumSize(Runtime.getRuntime().maxMemory())
+        .evictionListener(
+            (Map<String, String> key,
+                SnowflakeStreamingIngestClient client,
+                RemovalCause removalCause) -> {
+              streamingClientHandler.closeClient(client);
+              LOGGER.info(
+                  "Removed registered client {} due to {}",
+                  client.getName(),
+                  removalCause.toString());
+            })
+        .build(streamingClientHandler::createClient);
+  }
+
   /** ONLY FOR TESTING - to get a provider with injected properties */
   @VisibleForTesting
   public static StreamingClientProvider getStreamingClientProviderForTests(
-      SnowflakeStreamingIngestClient parameterEnabledClient,
-      StreamingClientHandler streamingClientHandler) {
-    return new StreamingClientProvider(parameterEnabledClient, streamingClientHandler);
+      StreamingClientHandler streamingClientHandler,
+      LoadingCache<Map<String, String>, SnowflakeStreamingIngestClient> registeredClients) {
+    return new StreamingClientProvider(streamingClientHandler, registeredClients);
   }
 
   /** ONLY FOR TESTING - private constructor to inject properties for testing */
   private StreamingClientProvider(
-      SnowflakeStreamingIngestClient parameterEnabledClient,
-      StreamingClientHandler streamingClientHandler) {
-    this();
-    this.parameterEnabledClient = parameterEnabledClient;
+      StreamingClientHandler streamingClientHandler,
+      LoadingCache<Map<String, String>, SnowflakeStreamingIngestClient> registeredClients) {
     this.streamingClientHandler = streamingClientHandler;
+    this.registeredClients = registeredClients;
   }
 
   private static final KCLogger LOGGER = new KCLogger(StreamingClientProvider.class.getName());
-  private SnowflakeStreamingIngestClient parameterEnabledClient;
   private StreamingClientHandler streamingClientHandler;
-  private Lock providerLock;
+  private LoadingCache<Map<String, String>, SnowflakeStreamingIngestClient> registeredClients;
```

Comment:
keying this cache on the connector config means that we are creating more clients. However I think this is better because connector configs would never be shared between clients. 

We can also use the Properties created convertConfigForStreamingClient which would be a subset of the connector configs. This would reduce the number of clients created, however I worry about accidentally leaking a config cross connector

Reply:
I dont think we should have key as a map in loadingcache,, 
How about we create a CacheKey object in this class with the required properties we care about? 
take a look at this for reference: https://github.com/snowflakedb/snowflake/blob/main/GlobalServices/modules/dbsec/authn-api/src/main/java/com/snowflake/security/ShiroPasswordHashCache.java#L148

Once you have a class as a key, you can then define your own equals and hashcode and that will guarantee to work as expected? 
For instance here you can add all properties as instance variable which you think is helpful for unique creation of a client. 

IMO,
ideally the connector name should be enough to create a client at a worker level. each task which belongs to that connector will get the same instance and all other connectors will create its own client..

Reply:
you're right, @sfc-gh-tzhang had the same concern so I've created a StreamingClientProperties class that will take in a connectorConfig and parse out the important properties passed into the client. The LoadingCache uses the StreamingClientProperties to determine if a new client needs to be created. This means connectors with the same configuration except different names or buffer flush thresholds would still use the same client